{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "\n",
    "%matplotlib notebook\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "from scipy import sparse\n",
    "\n",
    "import xml.etree.ElementTree as et\n",
    "import os\n",
    "\n",
    "import time\n",
    "import pickle\n",
    "import memory_profiler\n",
    "%load_ext memory_profiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from lxml import etree, objectify\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pyarrow.Table\n",
       "PMID: string\n",
       "ISSN: string\n",
       "Title: string\n",
       "ArticleTitle: string\n",
       "AbstractText: string\n",
       "Language: string\n",
       "Country: string\n",
       "MedlineTA: string\n",
       "ISSNLinking: string\n",
       "----\n",
       "PMID: [[\"18222646\",\"18222647\",\"18222648\",\"18222649\",\"18222652\",...,\"18254138\",\"18254137\",\"18254139\",\"18254140\",\"18254136\"]]\n",
       "ISSN: [[\"0896-8411\",\"0898-6568\",\"0920-9964\",\"0923-2508\",\"0924-977X\",...,\"0269-3879\",\"0960-7692\",\"0269-3879\",\"0269-3879\",\"1520-7552\"]]\n",
       "Title: [[\"Journal of autoimmunity\",\"Cellular signalling\",\"Schizophrenia research\",\"Research in microbiology\",\"European neuropsychopharmacology : the journal of the European College of Neuropsychopharmacology\",...,\"Biomedical chromatography : BMC\",\"Ultrasound in obstetrics & gynecology : the official journal of the International Society of Ultrasound in Obstetrics and Gynecology\",\"Biomedical chromatography : BMC\",\"Biomedical chromatography : BMC\",\"Diabetes/metabolism research and reviews\"]]\n",
       "ArticleTitle: [[\"Will hematopoietic stem cell transplantation cure human autoimmune diseases?\",\"Activation of c-Jun N-terminal kinase (JNK) by widely used specific p38 MAPK inhibitors SB202190 and SB203580: a MLK-3-MKK7-dependent mechanism.\",\"Neuropsychological predictors of functional outcome in Cognitive Behavioral Social Skills Training for older people with schizophrenia.\",\"El Tor and Calcutta CTXPhi precursors coexisting with intact CTXPhi copies in Vibrio cholerae O139 isolates.\",\"Hypothalamus volume in twin pairs discordant for schizophrenia.\",...,\"Determination of icariin in rat plasma by reverse-phase high-performance liquid chromatography after oral administration of a lipid-based suspension of Epimedium koreanum extract.\",\"Standardized views of the fetal heart using four-dimensional sonographic and tomographic imaging.\",\"Simple HPLC-UV determination of nucleosides and its application to the authentication of Cordyceps and its allies.\",\"Improved high-performance liquid chromatographic method for simultaneous determination of 12 cytotoxic caged xanthones in gamboges, a potential anticancer resin from Garcinia hanburyi.\",\"Hypouricemia linked to an overproduction of nitric oxide is an early marker of oxidative stress in female subjects with type 1 diabetes.\"]]\n",
       "AbstractText: [[\"Hematopoietic stem cell transplantation is becoming an accepted therapy for severe autoimmune diseases, and over 1,000 patients have been transplanted worldwide. Diseases include neurological (multiple sclerosis, others), rheumatological (systemic sclerosis, systemic lupus erythematosus, rheumatoid arthritis, vasculitis), haematological (aplastic anemia, single line immune mediated cytopenias) and others. The aim of this article is not to review the copious specific literature, but to analyze whether the up to now existing evidence satisfies the requirements of cure. Prospective randomized trials have been launched by the European Group of Blood and Marrow Transplantation (EBMT). Autologous transplantation, by far the most widely utilized because of its safety, has been shown to possess a powerful disease-arresting effect, but whether the attendant immune reconstitution (\"re-education\") will finally lead to cure is not demonstrated. The experience with allogeneic transplantation is too limited to draw even tentative conclusions. A Graft-versus-Autoimmunity effect has been ascertained both experimentally and clinically, but cure of autoimmunity by this procedure has not been demonstrated. Unexpected relapses in spite of full donor chimerism have been published. Further experience and studies are necessary.\",\"Mitogen-activated protein kinases (MAPKs) are key signaling molecules that respond to mitogenic stimulation or environmental stress, resulting in the expression of target proteins. c-Jun N-terminal kinase (JNK) and p38 MAPKs are activated by inflammatory cytokines or environmental stress. Specific p38 MAPK inhibitors, such as SB202190 or SB203580, are widely used to dissect p38 MAPK-related signal transduction mechanisms. While using SB202190 to inhibit p38 MAPK-related signaling, we observed that SB202190 treatment could activate JNK. Further experiments showed that treatment of cells with SB202190 could phosphorylate JNK and activating transcription factor 2 (ATF-2), and increased AP-1 DNA binding. Using multiple cell lines and primary endothelial cells, we demonstrated that specific p38 MAPK inhibitors SB202190 or SB203580 induces the activation of the JNK pathway. Further, using with RNA interference and kinase-inactive expression of intermediates of the JNK pathway, we demonstrated SB202190- or SB203580-induced JNK activation is dependent on the MLK-3-MKK4/MKK7-dependent signal transduction pathway. Finally, we demonstrate that treatment of cells with SB202190 or SB203580 induces the phosphorylation and activation of MLK3.\",\"Cognitive Behavioral Social Skills Training (CBSST) is a 24-session weekly group therapy intervention to improve functioning in people with schizophrenia. In our prior randomized clinical trial comparing treatment as usual (TAU) with TAU plus group CBSST (Granholm, E., McQuaid, J.R., McClure, F.S., Auslander, L., Perivoliotis, D., Pedrelli, P., Patterson, T., Jeste, D.V., 2005. A randomized controlled trial of cognitive behavioral social skills training for middle-aged and older outpatients with chronic schizophrenia. Am. J. Psychiatry 162, 520-529.), participants with schizophrenia in CBSST showed significantly better functional outcome than participants in TAU. The present study was a secondary analysis of neuropsychological predictors of functional outcome in our prior CBSST trial. We examined (1) whether neuropsychological impairment at baseline moderated functional outcome in CBSST relative to TAU, and (2) whether improvement in neuropsychological abilities mediated improvement in functional outcome in CBSST. Attention, verbal learning/memory, speed of processing, and executive functions were assessed at baseline, end of treatment, and 12-month follow-up. Greater severity of neuropsychological impairment at baseline predicted poorer functional outcome for both treatment groups (nonspecific predictor), but the interaction between severity of neuropsychological impairment and treatment group was not significant (no moderation). Effect sizes for the difference between treatment groups on functional outcome measures at 12-month follow-up were similar for participants with relatively mild (d=.44-.64) and severe (d=.29-.60) neuropsychological impairment. Results also did not support the hypothesis that improvement in neuropsychological abilities mediated improvement in functioning in CBSST. Adding CBSST to standard pharmacologic care, therefore, improved functioning relative to standard care alone, even for participants with severe neuropsychological impairment, and this improvement in functioning was not related to improvement in neuropsychological abilities in CBSST.\",\"Restriction fragment length polymorphism analyses of the array of CTXPhi prophages in strains CRC262 and CRC266 of Vibrio cholerae O139 revealed the presence of copies of complete CTXPhi and pre-CTXPhi prophages coexisting at a single chromosomal locus in each strain. Restriction pattern and comparative nucleotide sequence analysis revealed pre-CTXPhi precursors of both the El Tor and Calcutta lineages. Thus, we hypothesize that two precursor variants independently acquired cholera toxin genes and gave rise to the current El Tor and Calcutta CTXPhi prophages. We discuss the implications of these results in terms of the evolution and origin of the current diversity of CTXPhi prophages.\",\"Monozygotic and same-sex dizygotic twin pairs discordant for schizophrenia were compared with matched control twin pairs in order to disentangle genetic and environmental contributions to variation in hypothalamus volume. A decrease in hypothalamus volume was found in patients or discordant twin pairs compared to healthy controls which could be attributed to the decrease in total brain volume. Higher within-twin pair similarities in monozygotic compared to dizygotic twin pairs, suggests that hypothalamus volume might be partly genetically controlled.\",...,\"A simple, specific and sensitive method for quantitative determination of icariin in rat plasma using reverse-phase high-performance liquid chromatography with UV-detection was developed and applied to an animal study of a lipid-based suspension of the Epimedium koreanum extract in rats. Rutin was selected as the internal standard and methanol was found to be the best solvent for extraction of icariin from the plasma. Linearity was observed between 0.030 and 100 microg/mL (r > 0.99). The extraction recoveries of icariin and rutin were approximately 75 and 80%, respectively, in plasma. The intra- and inter-day coefficients of variation were less than 5%. The limit of detection was 6 ng/mL and the limit of quantification was 18 ng/mL.\",null,\"A simple HPLC-UV method combined with a simple extraction procedure of nucleosides (adenosine, cordycepin, 2'-deoxyadenosine, guanosine and uridine) was developed and applied to the authentication of Cordyceps and its allies. The separation was performed on a C(18) column by isocratic elution with acetonitrile-water, and UV detection at 260 nm. The amounts of adenosine, cordycepin, 2'-deoxyadenosine, guanosine and uridine in Cordyceps were 0.28-14.15, 0.006-6.36, 0.01-0.14, 0.68-14.79 and 0.19-20.29 mg/g, respectively. Among the nucleosides studied, cordycepin was characteristically included in Cordyceps militaris (L.) Link. (CM), which is one of key Cordyceps allies, and might be a good marker for authenticating CM. The ratio of nucleosides to adenosine contents in Cordyceps seemed to be a useful marker for authentication and quality control of Cordyceps.\",\"The potential anti-tumor activity of gamboges, a herbal medicine derived from Garcinia hanburyi, has increasingly gained the interest of scientist worldwide. The major components of gamboges are cytotoxic caged xanthones. In the present study, an improved HPLC method was developed to simultaneously quantify 12 caged xanthones, including three pairs of epimers and four pairs of trans-cis isomers, i.e. forbesione, isomorellic acid, morellic acid, R-30-hydroxygambogic acid, S-30-hydroxygambogic acid, isogambogenic acid, gambogenic acid, gambogellic acid, R-isogambogic acid, S-isogambogic acid, R-gambogic acid and S-gambogic acid. This method was validated to be sensitive, precise and accurate with limits of detection of 0.03-0.08 microg/mL, overall intra-day and inter-day variations less than 7.9% and overall recovery over 93.2%. The correlation coefficients (r(2)) of the calibration curves were higher than 0.995 for all analytes. The newly established method was successfully applied to reveal the difference in the chemical profiles and contents of these analytes in gamboges from different origins. It can be concluded that this method was not only an effective quality control method to ensure the safety and efficacy consistency of gamboges, but also a useful tool for screening and determining more potent cytotoxic xanthones with potential anticancer activity.\",\"Oxidative stress is already present in the early stages of type 1 diabetes. We conclude that the initial increase in oxidative stress could be linked to a reduction in plasmatic levels of uric acid, which is probably directly caused by an overproduction of NO.\"]]\n",
       "Language: [[\"eng\",\"eng\",\"eng\",\"eng\",\"eng\",...,\"eng\",\"eng\",\"eng\",\"eng\",\"eng\"]]\n",
       "Country: [[\"England\",\"England\",\"Netherlands\",\"France\",\"Netherlands\",...,\"England\",\"England\",\"England\",\"England\",\"England\"]]\n",
       "MedlineTA: [[\"J Autoimmun\",\"Cell Signal\",\"Schizophr Res\",\"Res Microbiol\",\"Eur Neuropsychopharmacol\",...,\"Biomed Chromatogr\",\"Ultrasound Obstet Gynecol\",\"Biomed Chromatogr\",\"Biomed Chromatogr\",\"Diabetes Metab Res Rev\"]]\n",
       "ISSNLinking: [[\"0896-8411\",\"0898-6568\",\"0920-9964\",\"0923-2508\",\"0924-977X\",...,\"0269-3879\",\"0960-7692\",\"0269-3879\",\"0269-3879\",\"1520-7552\"]]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "level = 0\n",
    "in_meta = False\n",
    "def article_yielder(fname, limit = float(\"inf\")):\n",
    "  r = gzip.open(fname, \"rb\")\n",
    "  for i, (event, elem) in enumerate(etree.iterparse(r, events=(\"end\", \"start\"))):\n",
    "    if event == \"start\" and elem.tag == \"MedlineCitation\":\n",
    "      d = {}\n",
    "    if i > limit:\n",
    "      break\n",
    "    if event == \"end\":\n",
    "      for k in ['PMID', 'AbstractText', 'ISSN', 'Language', 'MedlineTA', 'ISSNLinking', 'ArticleTitle', 'Country', 'Title']:\n",
    "        if elem.tag == k:\n",
    "          d[k] = elem.text      \n",
    "      if elem.tag == \"MedlineCitation\":\n",
    "        yield d\n",
    "\n",
    "import random        \n",
    "from io import BytesIO\n",
    "from pyarrow import Table\n",
    "from pathlib import Path\n",
    "\n",
    "def export_file(fn):\n",
    "  rows = []\n",
    "  for i, article in enumerate(article_yielder(fn)):\n",
    "    rows.append(article)\n",
    "  return Table.from_pylist(rows)\n",
    "files = [*Path(\"data\").glob(\"*.gz\")]\n",
    "random.shuffle(files)\n",
    "\n",
    "export_file(files[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function for importing each xml file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xml_import(xml_file):\n",
    "    \"\"\"Parses the following elements of each paper stored in the input XML file, and stores them in a pandas \n",
    "    DataFrame.\n",
    "    \n",
    "    Elements of each paper:\n",
    "    - PDMI (stored in <PDMI>).\n",
    "    - Title (stored in <ArticleTitle>).\n",
    "    - Abstract (stored in <AbstractText>).\n",
    "    - Language (stored in <Language>).\n",
    "    - Journal (stored in <Title>).\n",
    "    - Date (stored in <PubDate>).\n",
    "    \n",
    "    \n",
    "    \n",
    "    Details about information extracion: \n",
    "    - PDMI: If there is no tag <PDMI>, it will add 'no tag'. If there is more than one <PDMI>,\n",
    "    it will import only the first one. If <PDMI> contains no text, it will add '' (empty string).\n",
    "    \n",
    "    - Title: If there is no tag <ArticleTitle>, it will add 'no tag'. If there is more than one\n",
    "    <ArticleTitle>, it will import only the first one. If <ArticleTitle> contains no text, it will add '' (empty string).\n",
    "    \n",
    "    - Abstract: If there is no tag <Abstract> (parent of <AbstractText>), it will add '' (empty string).\n",
    "    If there is more than one <AbstractText> inside <Abstract>, it will combine them into one list.\n",
    "    If <AbstractText> contains no text, it will add '' (empty string).\n",
    "    If there is more than one <Abstract> or other tags containing <AbstractText>, like \n",
    "    <OtherAbstract>, it will not get text from them. \n",
    "    \n",
    "    - Language: If there is no tag <Language>, it will add 'no tag'. If there is more than one\n",
    "    <Language>, it will import only the first one. If <Language> contains no text, it will add '' (empty string).\n",
    "    \n",
    "    - Journal: If there is no tag <Title>, it will add 'no tag'. If there is more than one <Title>,\n",
    "    it will import only the first one. If <Title> contains no text, it will add '' (empty string).\n",
    "    \n",
    "    - Date: If there is no tag <PubDate>, it will add 'no tag'. It will combine all the <PubDate>'s childs' texts\n",
    "    into one (due to the assymetry of the date storage, sometimes with <Day>, <Month> and <Year>, other times \n",
    "    with <MedlineDate>). If <PubDate> contains no further childs, it will print ' '.\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    xtree = et.parse(xml_file)\n",
    "    xroot = xtree.getroot()\n",
    "\n",
    "    dicc={}\n",
    "\n",
    "    #PMID \n",
    "    ros=[]\n",
    "    for child1 in xroot:\n",
    "        for child2 in child1:\n",
    "            for element in child2.iter('MedlineCitation'):\n",
    "                tag=element.find('PMID')\n",
    "                if tag is None:\n",
    "                    ros.append(['no tag'])\n",
    "                else:\n",
    "                    res=[]\n",
    "                    if not tag.text :\n",
    "                        res.append('')\n",
    "                    else:\n",
    "                        res.append(tag.text)\n",
    "                    ros.append(res)\n",
    "\n",
    "\n",
    "    ros=[' '.join(ele) for ele in ros]\n",
    "    dicc['PMID']=ros\n",
    "\n",
    "    \n",
    "    #Title \n",
    "    ros=[]\n",
    "    for child1 in xroot:\n",
    "        for child2 in child1:\n",
    "            for child3 in child2:\n",
    "                for article in child3.iter('Article'):\n",
    "                    tag=article.find('ArticleTitle')\n",
    "                    if tag is None:\n",
    "                        ros.append(['no tag'])\n",
    "                    else:\n",
    "                        res=[]\n",
    "                        if not tag.text :\n",
    "                            res.append('')\n",
    "                        else:\n",
    "                            res.append(tag.text)\n",
    "                        ros.append(res)\n",
    "\n",
    "    ros=[' '.join(ele) for ele in ros]\n",
    "    dicc['Title']=ros\n",
    "\n",
    "\n",
    "    #Abstract \n",
    "    ros=[]\n",
    "    for child1 in xroot:\n",
    "        for child2 in child1:\n",
    "            for child3 in child2:\n",
    "                for article in child3.iter('Article'):\n",
    "                    tag=article.find('Abstract')\n",
    "                    if tag is None:\n",
    "                        ros.append([''])\n",
    "                    else:\n",
    "                        for child4 in child3:\n",
    "                            for elem in child4.iter('Abstract'):\n",
    "                                res=[]\n",
    "                                for AbstractText in elem.iter('AbstractText'):\n",
    "                                    if not AbstractText.text:\n",
    "                                        res.append('')\n",
    "                                    else:\n",
    "                                        res.append(AbstractText.text)\n",
    "                                ros.append(res)\n",
    "\n",
    "    ros=[' '.join(ele) for ele in ros]\n",
    "    dicc['AbstractText']=ros\n",
    "    \n",
    "    \n",
    "    #Language \n",
    "    ros=[]\n",
    "    for child1 in xroot:\n",
    "        for child2 in child1:\n",
    "            for child3 in child2:\n",
    "                for article in child3.iter('Article'):\n",
    "                    tag=article.find('Language')\n",
    "                    if tag is None:\n",
    "                        ros.append(['no tag'])\n",
    "                    else:\n",
    "                        res=[]\n",
    "                        if not tag.text :\n",
    "                            res.append('')\n",
    "                        else:\n",
    "                            res.append(tag.text)\n",
    "                        ros.append(res)\n",
    "\n",
    "    ros=[' '.join(ele) for ele in ros]\n",
    "    dicc['Language']=ros\n",
    "\n",
    "    \n",
    "    #Journal \n",
    "    ros=[]\n",
    "    for child1 in xroot:\n",
    "        for child2 in child1:\n",
    "            for child3 in child2:\n",
    "                for child4 in child3:\n",
    "                    for journal in child4.iter('Journal'):\n",
    "                        tag=journal.find('Title')\n",
    "                        if tag is None:\n",
    "                            ros.append(['no tag'])\n",
    "                        else:\n",
    "                            res=[]\n",
    "                            if not tag.text:\n",
    "                                res.append('')\n",
    "                            else:\n",
    "                                res.append(tag.text)\n",
    "                            ros.append(res)\n",
    "\n",
    "    ros=[' '.join(ele) for ele in ros]\n",
    "    dicc['Journal']=ros\n",
    "\n",
    "    \n",
    "    #Date\n",
    "    ros=[]\n",
    "    for child1 in xroot:\n",
    "        for child2 in child1:\n",
    "            for child3 in child2:\n",
    "                for child4 in child3:\n",
    "                    for child5 in child4:\n",
    "                        for JI in child5.iter('JournalIssue'):\n",
    "                            tag=JI.find('PubDate')\n",
    "                            if tag is None:\n",
    "                                ros.append(['no tag'])\n",
    "                            else:\n",
    "                                res=[]\n",
    "                                for elem in tag:\n",
    "                                    res.append(elem.text)\n",
    "                                ros.append(res)\n",
    "                                \n",
    "    ros=[' '.join(ele) for ele in ros]\n",
    "    dicc['Date']=ros\n",
    "    \n",
    "    out=pd.DataFrame.from_dict(dicc)\n",
    "    return out, dicc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function for importing all files from a directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_all_files(path, order_files=False):\n",
    "    \"\"\"\n",
    "    Imports all xml files from a directory into a combined dataframe using the function xml_import created above.\n",
    "    \n",
    "    Input:\n",
    "        - path : srt \"\" with the path of the directory with the files you want to import.\n",
    "        - order_files=False : if it is True it will print the order in which files are being imported.\n",
    "        \n",
    "    Output:\n",
    "        - final_df : Panda Dataframe with all the xml files from the directory imported and merged \n",
    "        together (concatenated in the order that they were in the directory (from up to down)).\n",
    "    \"\"\"\n",
    "    # name_files has the names of both .xml files and .gz.md5 files\n",
    "    name_files=os.listdir(path)\n",
    "    \n",
    "    # we select only the .xml files\n",
    "    len_filenames_list=map(len, name_files)\n",
    "    len_filenames=np.fromiter(len_filenames_list, dtype=np.int64,count=len(name_files))\n",
    "\n",
    "    name_files_array=np.array(name_files)\n",
    "    name_xml_files=name_files_array[len_filenames==17]\n",
    "    \n",
    "    # import\n",
    "    frame_all_df=[]\n",
    "    \n",
    "    for i in range(0,len(name_xml_files)):\n",
    "        path_file=path+name_xml_files[i]\n",
    "        if order_files==True:\n",
    "            print(name_xml_files[i])\n",
    "        df,dic=xml_import(str(path_file))\n",
    "        frame_all_df.append(df)\n",
    "\n",
    "    final_df=pd.concat(frame_all_df,ignore_index=True)\n",
    "    return final_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pubmed21n0038.xml\n",
      "pubmed21n0722.xml\n",
      "pubmed21n0814.xml\n",
      "pubmed21n0171.xml\n",
      "pubmed21n0124.xml\n",
      "pubmed21n0864.xml\n",
      "pubmed21n0876.xml\n",
      "pubmed21n0133.xml\n",
      "pubmed21n0539.xml\n",
      "pubmed21n0119.xml\n",
      "pubmed21n0056.xml\n",
      "pubmed21n0497.xml\n",
      "pubmed21n0821.xml\n",
      "pubmed21n0241.xml\n",
      "pubmed21n0986.xml\n",
      "pubmed21n0769.xml\n",
      "pubmed21n0857.xml\n",
      "pubmed21n0073.xml\n",
      "pubmed21n0479.xml\n",
      "pubmed21n0044.xml\n",
      "pubmed21n0208.xml\n",
      "pubmed21n0071.xml\n",
      "pubmed21n0014.xml\n",
      "pubmed21n0795.xml\n",
      "pubmed21n0074.xml\n",
      "pubmed21n0662.xml\n",
      "pubmed21n0543.xml\n",
      "pubmed21n0555.xml\n",
      "pubmed21n0048.xml\n",
      "pubmed21n0512.xml\n",
      "pubmed21n0152.xml\n",
      "pubmed21n0078.xml\n",
      "pubmed21n0093.xml\n",
      "pubmed21n1051.xml\n",
      "pubmed21n0304.xml\n",
      "pubmed21n0172.xml\n",
      "pubmed21n0361.xml\n",
      "pubmed21n0831.xml\n",
      "pubmed21n0376.xml\n",
      "pubmed21n0203.xml\n",
      "pubmed21n0885.xml\n",
      "pubmed21n0740.xml\n",
      "pubmed21n0154.xml\n",
      "pubmed21n1005.xml\n",
      "pubmed21n0398.xml\n",
      "pubmed21n0079.xml\n",
      "pubmed21n1011.xml\n",
      "pubmed21n0049.xml\n",
      "pubmed21n0929.xml\n",
      "pubmed21n0404.xml\n",
      "pubmed21n0887.xml\n",
      "pubmed21n0911.xml\n",
      "pubmed21n0668.xml\n",
      "pubmed21n0231.xml\n",
      "pubmed21n0013.xml\n",
      "pubmed21n0283.xml\n",
      "pubmed21n0837.xml\n",
      "pubmed21n0290.xml\n",
      "pubmed21n0549.xml\n",
      "pubmed21n0832.xml\n",
      "pubmed21n0884.xml\n",
      "pubmed21n0175.xml\n",
      "pubmed21n0665.xml\n",
      "pubmed21n0980.xml\n",
      "pubmed21n0474.xml\n",
      "pubmed21n0103.xml\n",
      "pubmed21n0352.xml\n",
      "pubmed21n0503.xml\n",
      "pubmed21n0766.xml\n",
      "pubmed21n0242.xml\n",
      "pubmed21n0216.xml\n",
      "pubmed21n0867.xml\n",
      "pubmed21n0550.xml\n",
      "pubmed21n0475.xml\n",
      "pubmed21n0303.xml\n",
      "pubmed21n0045.xml\n",
      "pubmed21n0422.xml\n",
      "pubmed21n0554.xml\n",
      "pubmed21n0200.xml\n",
      "pubmed21n0443.xml\n",
      "pubmed21n0772.xml\n",
      "pubmed21n0105.xml\n",
      "pubmed21n0002.xml\n",
      "pubmed21n0392.xml\n",
      "pubmed21n0173.xml\n",
      "pubmed21n0993.xml\n",
      "pubmed21n0396.xml\n",
      "pubmed21n0732.xml\n",
      "pubmed21n0725.xml\n",
      "pubmed21n1058.xml\n",
      "pubmed21n0839.xml\n",
      "pubmed21n0067.xml\n",
      "pubmed21n0631.xml\n",
      "pubmed21n0302.xml\n",
      "pubmed21n0449.xml\n",
      "pubmed21n0137.xml\n",
      "pubmed21n0658.xml\n",
      "pubmed21n1020.xml\n",
      "pubmed21n1054.xml\n",
      "pubmed21n0687.xml\n",
      "pubmed21n0612.xml\n",
      "pubmed21n0027.xml\n",
      "pubmed21n0810.xml\n",
      "pubmed21n0205.xml\n",
      "pubmed21n0031.xml\n",
      "pubmed21n0985.xml\n",
      "pubmed21n0366.xml\n",
      "pubmed21n0801.xml\n",
      "pubmed21n0164.xml\n",
      "pubmed21n0782.xml\n",
      "pubmed21n0926.xml\n",
      "pubmed21n0170.xml\n",
      "pubmed21n0359.xml\n",
      "pubmed21n0347.xml\n",
      "pubmed21n0899.xml\n",
      "pubmed21n0260.xml\n",
      "pubmed21n0660.xml\n",
      "pubmed21n0788.xml\n",
      "pubmed21n0019.xml\n",
      "pubmed21n0506.xml\n",
      "pubmed21n0115.xml\n",
      "pubmed21n0464.xml\n",
      "pubmed21n0652.xml\n",
      "pubmed21n0082.xml\n",
      "pubmed21n0519.xml\n",
      "pubmed21n0854.xml\n",
      "pubmed21n0733.xml\n",
      "pubmed21n0809.xml\n",
      "pubmed21n0369.xml\n",
      "pubmed21n0485.xml\n",
      "pubmed21n0322.xml\n",
      "pubmed21n0958.xml\n",
      "pubmed21n0808.xml\n",
      "pubmed21n0623.xml\n",
      "pubmed21n0542.xml\n",
      "pubmed21n0495.xml\n",
      "pubmed21n0022.xml\n",
      "pubmed21n0391.xml\n",
      "pubmed21n1061.xml\n",
      "pubmed21n0537.xml\n",
      "pubmed21n0356.xml\n",
      "pubmed21n0998.xml\n",
      "pubmed21n0341.xml\n",
      "pubmed21n0780.xml\n",
      "pubmed21n0778.xml\n",
      "pubmed21n0869.xml\n",
      "pubmed21n0462.xml\n",
      "pubmed21n0129.xml\n",
      "pubmed21n0737.xml\n",
      "pubmed21n0728.xml\n",
      "pubmed21n0092.xml\n",
      "pubmed21n0295.xml\n",
      "pubmed21n0969.xml\n",
      "pubmed21n0674.xml\n",
      "pubmed21n0087.xml\n",
      "pubmed21n0646.xml\n",
      "pubmed21n0223.xml\n",
      "pubmed21n0088.xml\n",
      "pubmed21n0804.xml\n",
      "pubmed21n0224.xml\n",
      "pubmed21n1029.xml\n",
      "pubmed21n0128.xml\n",
      "pubmed21n0528.xml\n",
      "pubmed21n0987.xml\n",
      "pubmed21n0402.xml\n",
      "pubmed21n0317.xml\n",
      "pubmed21n0028.xml\n",
      "pubmed21n0197.xml\n",
      "pubmed21n0308.xml\n",
      "pubmed21n0293.xml\n",
      "pubmed21n0960.xml\n",
      "pubmed21n0012.xml\n",
      "pubmed21n1013.xml\n",
      "pubmed21n0117.xml\n",
      "pubmed21n0393.xml\n",
      "pubmed21n0976.xml\n",
      "pubmed21n0381.xml\n",
      "pubmed21n0099.xml\n",
      "pubmed21n0195.xml\n",
      "pubmed21n0597.xml\n",
      "pubmed21n0297.xml\n",
      "pubmed21n0496.xml\n",
      "pubmed21n0708.xml\n",
      "pubmed21n0703.xml\n",
      "pubmed21n0315.xml\n",
      "pubmed21n0977.xml\n",
      "pubmed21n0914.xml\n",
      "pubmed21n0063.xml\n",
      "pubmed21n0746.xml\n",
      "pubmed21n0803.xml\n",
      "pubmed21n0997.xml\n",
      "pubmed21n0953.xml\n",
      "pubmed21n0925.xml\n",
      "pubmed21n0132.xml\n",
      "pubmed21n0276.xml\n",
      "pubmed21n0630.xml\n",
      "pubmed21n0018.xml\n",
      "pubmed21n0106.xml\n",
      "pubmed21n0797.xml\n",
      "pubmed21n0069.xml\n",
      "pubmed21n0967.xml\n",
      "pubmed21n0942.xml\n",
      "pubmed21n0165.xml\n",
      "pubmed21n0182.xml\n",
      "pubmed21n0619.xml\n",
      "pubmed21n0822.xml\n",
      "pubmed21n0368.xml\n",
      "pubmed21n0050.xml\n",
      "pubmed21n0228.xml\n",
      "pubmed21n0964.xml\n",
      "pubmed21n0779.xml\n",
      "pubmed21n0371.xml\n",
      "pubmed21n0812.xml\n",
      "pubmed21n0709.xml\n",
      "pubmed21n0158.xml\n",
      "pubmed21n0793.xml\n",
      "pubmed21n0472.xml\n",
      "pubmed21n0146.xml\n",
      "pubmed21n0902.xml\n",
      "pubmed21n0100.xml\n",
      "pubmed21n0544.xml\n",
      "pubmed21n0536.xml\n",
      "pubmed21n0486.xml\n",
      "pubmed21n0702.xml\n",
      "pubmed21n0363.xml\n",
      "pubmed21n0928.xml\n",
      "pubmed21n0441.xml\n",
      "pubmed21n0760.xml\n",
      "pubmed21n0177.xml\n",
      "pubmed21n0367.xml\n",
      "pubmed21n0336.xml\n",
      "pubmed21n0135.xml\n",
      "pubmed21n0162.xml\n",
      "pubmed21n0939.xml\n",
      "pubmed21n0377.xml\n",
      "pubmed21n0681.xml\n",
      "pubmed21n0686.xml\n",
      "pubmed21n0450.xml\n",
      "pubmed21n0238.xml\n",
      "pubmed21n0326.xml\n",
      "pubmed21n0345.xml\n",
      "pubmed21n0094.xml\n",
      "pubmed21n0871.xml\n",
      "pubmed21n0227.xml\n",
      "pubmed21n0920.xml\n",
      "pubmed21n0773.xml\n",
      "pubmed21n0206.xml\n",
      "pubmed21n0024.xml\n",
      "pubmed21n0918.xml\n",
      "pubmed21n1030.xml\n",
      "pubmed21n0025.xml\n",
      "pubmed21n0330.xml\n",
      "pubmed21n0041.xml\n",
      "pubmed21n0730.xml\n",
      "pubmed21n1024.xml\n",
      "pubmed21n0246.xml\n",
      "pubmed21n0529.xml\n",
      "pubmed21n0842.xml\n",
      "pubmed21n0118.xml\n",
      "pubmed21n0594.xml\n",
      "pubmed21n0310.xml\n",
      "pubmed21n0346.xml\n",
      "pubmed21n0640.xml\n",
      "pubmed21n0971.xml\n",
      "pubmed21n0683.xml\n",
      "pubmed21n0207.xml\n",
      "pubmed21n0845.xml\n",
      "pubmed21n0273.xml\n",
      "pubmed21n0932.xml\n",
      "pubmed21n0641.xml\n",
      "pubmed21n0786.xml\n",
      "pubmed21n0226.xml\n",
      "pubmed21n0190.xml\n",
      "pubmed21n0001.xml\n",
      "pubmed21n0714.xml\n",
      "pubmed21n1049.xml\n",
      "pubmed21n0225.xml\n",
      "pubmed21n0489.xml\n",
      "pubmed21n0859.xml\n",
      "pubmed21n0840.xml\n",
      "pubmed21n0666.xml\n",
      "pubmed21n0951.xml\n",
      "pubmed21n0734.xml\n",
      "pubmed21n0524.xml\n",
      "pubmed21n1062.xml\n",
      "pubmed21n0613.xml\n",
      "pubmed21n0339.xml\n",
      "pubmed21n0944.xml\n",
      "pubmed21n0627.xml\n",
      "pubmed21n0882.xml\n",
      "pubmed21n0349.xml\n",
      "pubmed21n0416.xml\n",
      "pubmed21n0150.xml\n",
      "pubmed21n0400.xml\n",
      "pubmed21n0763.xml\n",
      "pubmed21n0584.xml\n",
      "pubmed21n0230.xml\n",
      "pubmed21n0930.xml\n",
      "pubmed21n0247.xml\n",
      "pubmed21n0789.xml\n",
      "pubmed21n0664.xml\n",
      "pubmed21n0006.xml\n",
      "pubmed21n0824.xml\n",
      "pubmed21n0202.xml\n",
      "pubmed21n1004.xml\n",
      "pubmed21n0995.xml\n",
      "pubmed21n0695.xml\n",
      "pubmed21n1048.xml\n",
      "pubmed21n0642.xml\n",
      "pubmed21n0583.xml\n",
      "pubmed21n0974.xml\n",
      "pubmed21n0196.xml\n",
      "pubmed21n0973.xml\n",
      "pubmed21n0603.xml\n",
      "pubmed21n0343.xml\n",
      "pubmed21n0718.xml\n",
      "pubmed21n0581.xml\n",
      "pubmed21n0813.xml\n",
      "pubmed21n1006.xml\n",
      "pubmed21n0429.xml\n",
      "pubmed21n0807.xml\n",
      "pubmed21n0729.xml\n",
      "pubmed21n0390.xml\n",
      "pubmed21n0278.xml\n",
      "pubmed21n0739.xml\n",
      "pubmed21n0384.xml\n",
      "pubmed21n0186.xml\n",
      "pubmed21n0328.xml\n",
      "pubmed21n0426.xml\n",
      "pubmed21n0005.xml\n",
      "pubmed21n0604.xml\n",
      "pubmed21n0700.xml\n",
      "pubmed21n0333.xml\n",
      "pubmed21n0270.xml\n",
      "pubmed21n0916.xml\n",
      "pubmed21n0575.xml\n",
      "pubmed21n0878.xml\n",
      "pubmed21n1012.xml\n",
      "pubmed21n0557.xml\n",
      "pubmed21n1046.xml\n",
      "pubmed21n0427.xml\n",
      "pubmed21n0615.xml\n",
      "pubmed21n0811.xml\n",
      "pubmed21n0704.xml\n",
      "pubmed21n0562.xml\n",
      "pubmed21n0102.xml\n",
      "pubmed21n0941.xml\n",
      "pubmed21n0187.xml\n",
      "pubmed21n0251.xml\n",
      "pubmed21n0903.xml\n",
      "pubmed21n0794.xml\n",
      "pubmed21n0711.xml\n",
      "pubmed21n0095.xml\n",
      "pubmed21n0149.xml\n",
      "pubmed21n0855.xml\n",
      "pubmed21n0858.xml\n",
      "pubmed21n0240.xml\n",
      "pubmed21n0991.xml\n",
      "pubmed21n0558.xml\n",
      "pubmed21n0523.xml\n",
      "pubmed21n0957.xml\n",
      "pubmed21n0534.xml\n",
      "pubmed21n0735.xml\n",
      "pubmed21n0915.xml\n",
      "pubmed21n0406.xml\n",
      "pubmed21n0830.xml\n",
      "pubmed21n0547.xml\n",
      "pubmed21n0204.xml\n",
      "pubmed21n0312.xml\n",
      "pubmed21n0651.xml\n",
      "pubmed21n0008.xml\n",
      "pubmed21n0897.xml\n",
      "pubmed21n0643.xml\n",
      "pubmed21n1027.xml\n",
      "pubmed21n1007.xml\n",
      "pubmed21n0478.xml\n",
      "pubmed21n0752.xml\n",
      "pubmed21n0841.xml\n",
      "pubmed21n0468.xml\n",
      "pubmed21n0344.xml\n",
      "pubmed21n1034.xml\n",
      "pubmed21n0234.xml\n",
      "pubmed21n0819.xml\n",
      "pubmed21n1016.xml\n",
      "pubmed21n0896.xml\n",
      "pubmed21n0166.xml\n",
      "pubmed21n0743.xml\n",
      "pubmed21n0209.xml\n",
      "pubmed21n0338.xml\n",
      "pubmed21n0890.xml\n",
      "pubmed21n0965.xml\n",
      "pubmed21n0636.xml\n",
      "pubmed21n0984.xml\n",
      "pubmed21n0244.xml\n",
      "pubmed21n0717.xml\n",
      "pubmed21n0790.xml\n",
      "pubmed21n0081.xml\n",
      "pubmed21n0280.xml\n",
      "pubmed21n0125.xml\n",
      "pubmed21n0940.xml\n",
      "pubmed21n0191.xml\n",
      "pubmed21n0378.xml\n",
      "pubmed21n0963.xml\n",
      "pubmed21n0222.xml\n",
      "pubmed21n0756.xml\n",
      "pubmed21n0836.xml\n",
      "pubmed21n0120.xml\n",
      "pubmed21n0777.xml\n",
      "pubmed21n0448.xml\n",
      "pubmed21n0852.xml\n",
      "pubmed21n0324.xml\n",
      "pubmed21n0292.xml\n",
      "pubmed21n0499.xml\n",
      "pubmed21n0679.xml\n",
      "pubmed21n0083.xml\n",
      "pubmed21n0996.xml\n",
      "pubmed21n0143.xml\n",
      "pubmed21n0034.xml\n",
      "pubmed21n0284.xml\n",
      "pubmed21n0937.xml\n",
      "pubmed21n0250.xml\n",
      "pubmed21n0650.xml\n",
      "pubmed21n0802.xml\n",
      "pubmed21n0502.xml\n",
      "pubmed21n0582.xml\n",
      "pubmed21n0259.xml\n",
      "pubmed21n0571.xml\n",
      "pubmed21n0599.xml\n",
      "pubmed21n0047.xml\n",
      "pubmed21n0561.xml\n",
      "pubmed21n0026.xml\n",
      "pubmed21n0313.xml\n",
      "pubmed21n0245.xml\n",
      "pubmed21n0560.xml\n",
      "pubmed21n0364.xml\n",
      "pubmed21n0881.xml\n",
      "pubmed21n0373.xml\n",
      "pubmed21n0514.xml\n",
      "pubmed21n0654.xml\n",
      "pubmed21n0917.xml\n",
      "pubmed21n0353.xml\n",
      "pubmed21n0155.xml\n",
      "pubmed21n0872.xml\n",
      "pubmed21n0927.xml\n",
      "pubmed21n0147.xml\n",
      "pubmed21n0553.xml\n",
      "pubmed21n0327.xml\n",
      "pubmed21n0329.xml\n",
      "pubmed21n0933.xml\n",
      "pubmed21n0112.xml\n",
      "pubmed21n0425.xml\n",
      "pubmed21n0649.xml\n",
      "pubmed21n0413.xml\n",
      "pubmed21n0589.xml\n",
      "pubmed21n0762.xml\n",
      "pubmed21n0277.xml\n",
      "pubmed21n0471.xml\n",
      "pubmed21n0268.xml\n",
      "pubmed21n0385.xml\n",
      "pubmed21n0624.xml\n",
      "pubmed21n0538.xml\n",
      "pubmed21n0744.xml\n",
      "pubmed21n0274.xml\n",
      "pubmed21n0127.xml\n",
      "pubmed21n0440.xml\n",
      "pubmed21n0817.xml\n",
      "pubmed21n0741.xml\n",
      "pubmed21n0690.xml\n",
      "pubmed21n0483.xml\n",
      "pubmed21n0053.xml\n",
      "pubmed21n0068.xml\n",
      "pubmed21n1044.xml\n",
      "pubmed21n0248.xml\n",
      "pubmed21n0596.xml\n",
      "pubmed21n0482.xml\n",
      "pubmed21n0070.xml\n",
      "pubmed21n0892.xml\n",
      "pubmed21n0491.xml\n",
      "pubmed21n0816.xml\n",
      "pubmed21n0913.xml\n",
      "pubmed21n0296.xml\n",
      "pubmed21n0676.xml\n",
      "pubmed21n0950.xml\n",
      "pubmed21n0895.xml\n",
      "pubmed21n0723.xml\n",
      "pubmed21n0983.xml\n",
      "pubmed21n0057.xml\n",
      "pubmed21n0463.xml\n",
      "pubmed21n0109.xml\n",
      "pubmed21n0340.xml\n",
      "pubmed21n0291.xml\n",
      "pubmed21n0688.xml\n",
      "pubmed21n0515.xml\n",
      "pubmed21n0061.xml\n",
      "pubmed21n0692.xml\n",
      "pubmed21n0431.xml\n",
      "pubmed21n0875.xml\n",
      "pubmed21n0428.xml\n",
      "pubmed21n0321.xml\n",
      "pubmed21n0286.xml\n",
      "pubmed21n0184.xml\n",
      "pubmed21n0264.xml\n",
      "pubmed21n0607.xml\n",
      "pubmed21n0705.xml\n",
      "pubmed21n0606.xml\n",
      "pubmed21n0738.xml\n",
      "pubmed21n0370.xml\n",
      "pubmed21n0163.xml\n",
      "pubmed21n0098.xml\n",
      "pubmed21n0707.xml\n",
      "pubmed21n0966.xml\n",
      "pubmed21n0586.xml\n",
      "pubmed21n0255.xml\n",
      "pubmed21n1019.xml\n",
      "pubmed21n0865.xml\n",
      "pubmed21n0210.xml\n",
      "pubmed21n0909.xml\n",
      "pubmed21n1037.xml\n",
      "pubmed21n0616.xml\n",
      "pubmed21n0629.xml\n",
      "pubmed21n0721.xml\n",
      "pubmed21n0551.xml\n",
      "pubmed21n0153.xml\n",
      "pubmed21n0874.xml\n",
      "pubmed21n0745.xml\n",
      "pubmed21n0860.xml\n",
      "pubmed21n0931.xml\n",
      "pubmed21n0675.xml\n",
      "pubmed21n0097.xml\n",
      "pubmed21n0040.xml\n",
      "pubmed21n0706.xml\n",
      "pubmed21n0358.xml\n",
      "pubmed21n0829.xml\n",
      "pubmed21n0439.xml\n",
      "pubmed21n1000.xml\n",
      "pubmed21n0693.xml\n",
      "pubmed21n0843.xml\n",
      "pubmed21n0633.xml\n",
      "pubmed21n0301.xml\n",
      "pubmed21n0691.xml\n",
      "pubmed21n0806.xml\n",
      "pubmed21n0578.xml\n",
      "pubmed21n0060.xml\n",
      "pubmed21n0853.xml\n",
      "pubmed21n0954.xml\n",
      "pubmed21n0588.xml\n",
      "pubmed21n0617.xml\n",
      "pubmed21n0217.xml\n",
      "pubmed21n0521.xml\n",
      "pubmed21n0130.xml\n",
      "pubmed21n0900.xml\n",
      "pubmed21n0775.xml\n",
      "pubmed21n0906.xml\n",
      "pubmed21n0458.xml\n",
      "pubmed21n0066.xml\n",
      "pubmed21n0380.xml\n",
      "pubmed21n0279.xml\n",
      "pubmed21n0355.xml\n",
      "pubmed21n0590.xml\n",
      "pubmed21n0111.xml\n",
      "pubmed21n0375.xml\n",
      "pubmed21n0791.xml\n",
      "pubmed21n0844.xml\n",
      "pubmed21n0530.xml\n",
      "pubmed21n0199.xml\n",
      "pubmed21n0566.xml\n",
      "pubmed21n0696.xml\n",
      "pubmed21n1045.xml\n",
      "pubmed21n0035.xml\n",
      "pubmed21n0444.xml\n",
      "pubmed21n0600.xml\n",
      "pubmed21n0667.xml\n",
      "pubmed21n0176.xml\n",
      "pubmed21n0213.xml\n",
      "pubmed21n0595.xml\n",
      "pubmed21n0201.xml\n",
      "pubmed21n0220.xml\n",
      "pubmed21n0469.xml\n",
      "pubmed21n0476.xml\n",
      "pubmed21n1023.xml\n",
      "pubmed21n0653.xml\n",
      "pubmed21n0386.xml\n",
      "pubmed21n0579.xml\n",
      "pubmed21n0614.xml\n",
      "pubmed21n0348.xml\n",
      "pubmed21n0879.xml\n",
      "pubmed21n0713.xml\n",
      "pubmed21n0682.xml\n",
      "pubmed21n0628.xml\n",
      "pubmed21n0540.xml\n",
      "pubmed21n1038.xml\n",
      "pubmed21n0838.xml\n",
      "pubmed21n0010.xml\n",
      "pubmed21n0559.xml\n",
      "pubmed21n0272.xml\n",
      "pubmed21n0023.xml\n",
      "pubmed21n0635.xml\n",
      "pubmed21n0131.xml\n",
      "pubmed21n0156.xml\n",
      "pubmed21n0754.xml\n",
      "pubmed21n0007.xml\n",
      "pubmed21n0608.xml\n",
      "pubmed21n0805.xml\n",
      "pubmed21n0139.xml\n",
      "pubmed21n0663.xml\n",
      "pubmed21n0758.xml\n",
      "pubmed21n0508.xml\n",
      "pubmed21n0421.xml\n",
      "pubmed21n0437.xml\n",
      "pubmed21n1002.xml\n",
      "pubmed21n0211.xml\n",
      "pubmed21n0701.xml\n",
      "pubmed21n0771.xml\n",
      "pubmed21n1022.xml\n",
      "pubmed21n0868.xml\n",
      "pubmed21n0395.xml\n",
      "pubmed21n0952.xml\n",
      "pubmed21n0388.xml\n",
      "pubmed21n0305.xml\n",
      "pubmed21n1010.xml\n",
      "pubmed21n0994.xml\n",
      "pubmed21n0504.xml\n",
      "pubmed21n0972.xml\n",
      "pubmed21n0316.xml\n",
      "pubmed21n0430.xml\n",
      "pubmed21n0084.xml\n",
      "pubmed21n1008.xml\n",
      "pubmed21n0043.xml\n",
      "pubmed21n0412.xml\n",
      "pubmed21n0481.xml\n",
      "pubmed21n0372.xml\n",
      "pubmed21n0862.xml\n",
      "pubmed21n0033.xml\n",
      "pubmed21n0168.xml\n",
      "pubmed21n1001.xml\n",
      "pubmed21n0281.xml\n",
      "pubmed21n0042.xml\n",
      "pubmed21n0531.xml\n",
      "pubmed21n0655.xml\n",
      "pubmed21n0116.xml\n",
      "pubmed21n0275.xml\n",
      "pubmed21n0294.xml\n",
      "pubmed21n0473.xml\n",
      "pubmed21n0221.xml\n",
      "pubmed21n0638.xml\n",
      "pubmed21n0114.xml\n",
      "pubmed21n0510.xml\n",
      "pubmed21n0436.xml\n",
      "pubmed21n0017.xml\n",
      "pubmed21n0287.xml\n",
      "pubmed21n0403.xml\n",
      "pubmed21n0126.xml\n",
      "pubmed21n0861.xml\n",
      "pubmed21n0401.xml\n",
      "pubmed21n0580.xml\n",
      "pubmed21n0492.xml\n",
      "pubmed21n0252.xml\n",
      "pubmed21n0546.xml\n",
      "pubmed21n0907.xml\n",
      "pubmed21n0678.xml\n",
      "pubmed21n0104.xml\n",
      "pubmed21n1018.xml\n",
      "pubmed21n0815.xml\n",
      "pubmed21n0432.xml\n",
      "pubmed21n0160.xml\n",
      "pubmed21n0335.xml\n",
      "pubmed21n0765.xml\n",
      "pubmed21n0142.xml\n",
      "pubmed21n0289.xml\n",
      "pubmed21n0856.xml\n",
      "pubmed21n0891.xml\n",
      "pubmed21n0029.xml\n",
      "pubmed21n0866.xml\n",
      "pubmed21n0625.xml\n",
      "pubmed21n0568.xml\n",
      "pubmed21n1036.xml\n",
      "pubmed21n0307.xml\n",
      "pubmed21n0015.xml\n",
      "pubmed21n0552.xml\n",
      "pubmed21n0189.xml\n",
      "pubmed21n0314.xml\n",
      "pubmed21n0731.xml\n",
      "pubmed21n0908.xml\n",
      "pubmed21n0318.xml\n",
      "pubmed21n0572.xml\n",
      "pubmed21n0905.xml\n",
      "pubmed21n1026.xml\n",
      "pubmed21n0934.xml\n",
      "pubmed21n0178.xml\n",
      "pubmed21n0062.xml\n",
      "pubmed21n0727.xml\n",
      "pubmed21n0922.xml\n",
      "pubmed21n0894.xml\n",
      "pubmed21n0459.xml\n",
      "pubmed21n0901.xml\n",
      "pubmed21n0526.xml\n",
      "pubmed21n0477.xml\n",
      "pubmed21n0699.xml\n",
      "pubmed21n0397.xml\n",
      "pubmed21n0254.xml\n",
      "pubmed21n0935.xml\n",
      "pubmed21n0055.xml\n",
      "pubmed21n0567.xml\n",
      "pubmed21n0169.xml\n",
      "pubmed21n0237.xml\n",
      "pubmed21n0256.xml\n",
      "pubmed21n0113.xml\n",
      "pubmed21n0320.xml\n",
      "pubmed21n0077.xml\n",
      "pubmed21n0415.xml\n",
      "pubmed21n0955.xml\n",
      "pubmed21n0846.xml\n",
      "pubmed21n0632.xml\n",
      "pubmed21n0520.xml\n",
      "pubmed21n1042.xml\n",
      "pubmed21n0350.xml\n",
      "pubmed21n0750.xml\n",
      "pubmed21n0420.xml\n",
      "pubmed21n0949.xml\n",
      "pubmed21n0564.xml\n",
      "pubmed21n0522.xml\n",
      "pubmed21n0533.xml\n",
      "pubmed21n0141.xml\n",
      "pubmed21n0285.xml\n",
      "pubmed21n0877.xml\n",
      "pubmed21n0818.xml\n",
      "pubmed21n0618.xml\n",
      "pubmed21n0144.xml\n",
      "pubmed21n0507.xml\n",
      "pubmed21n0004.xml\n",
      "pubmed21n0183.xml\n",
      "pubmed21n0036.xml\n",
      "pubmed21n1039.xml\n",
      "pubmed21n1050.xml\n",
      "pubmed21n0517.xml\n",
      "pubmed21n1028.xml\n",
      "pubmed21n0982.xml\n",
      "pubmed21n0138.xml\n",
      "pubmed21n0194.xml\n",
      "pubmed21n0229.xml\n",
      "pubmed21n0455.xml\n",
      "pubmed21n0759.xml\n",
      "pubmed21n0669.xml\n",
      "pubmed21n0362.xml\n",
      "pubmed21n0161.xml\n",
      "pubmed21n0360.xml\n",
      "pubmed21n0767.xml\n",
      "pubmed21n0748.xml\n",
      "pubmed21n0089.xml\n",
      "pubmed21n0179.xml\n",
      "pubmed21n0466.xml\n",
      "pubmed21n0054.xml\n",
      "pubmed21n0787.xml\n",
      "pubmed21n0453.xml\n",
      "pubmed21n0188.xml\n",
      "pubmed21n0319.xml\n",
      "pubmed21n0946.xml\n",
      "pubmed21n0389.xml\n",
      "pubmed21n0484.xml\n",
      "pubmed21n0418.xml\n",
      "pubmed21n0989.xml\n",
      "pubmed21n0334.xml\n",
      "pubmed21n0446.xml\n",
      "pubmed21n0409.xml\n",
      "pubmed21n0576.xml\n",
      "pubmed21n0621.xml\n",
      "pubmed21n0605.xml\n",
      "pubmed21n0783.xml\n",
      "pubmed21n0671.xml\n",
      "pubmed21n0956.xml\n",
      "pubmed21n0945.xml\n",
      "pubmed21n0021.xml\n",
      "pubmed21n0140.xml\n",
      "pubmed21n0968.xml\n",
      "pubmed21n0298.xml\n",
      "pubmed21n0828.xml\n",
      "pubmed21n0761.xml\n",
      "pubmed21n0332.xml\n",
      "pubmed21n0423.xml\n",
      "pubmed21n0424.xml\n",
      "pubmed21n0365.xml\n",
      "pubmed21n0609.xml\n",
      "pubmed21n0622.xml\n",
      "pubmed21n0685.xml\n",
      "pubmed21n0569.xml\n",
      "pubmed21n0747.xml\n",
      "pubmed21n0309.xml\n",
      "pubmed21n0672.xml\n",
      "pubmed21n0331.xml\n",
      "pubmed21n0212.xml\n",
      "pubmed21n0626.xml\n",
      "pubmed21n0342.xml\n",
      "pubmed21n0525.xml\n",
      "pubmed21n1032.xml\n",
      "pubmed21n0085.xml\n",
      "pubmed21n0030.xml\n",
      "pubmed21n0720.xml\n",
      "pubmed21n0670.xml\n",
      "pubmed21n1015.xml\n",
      "pubmed21n0947.xml\n",
      "pubmed21n1009.xml\n",
      "pubmed21n0382.xml\n",
      "pubmed21n0351.xml\n",
      "pubmed21n0535.xml\n",
      "pubmed21n1021.xml\n",
      "pubmed21n0611.xml\n",
      "pubmed21n0585.xml\n",
      "pubmed21n0962.xml\n",
      "pubmed21n0505.xml\n",
      "pubmed21n0121.xml\n",
      "pubmed21n0796.xml\n",
      "pubmed21n1041.xml\n",
      "pubmed21n0610.xml\n",
      "pubmed21n0039.xml\n",
      "pubmed21n0032.xml\n",
      "pubmed21n1043.xml\n",
      "pubmed21n0716.xml\n",
      "pubmed21n0556.xml\n",
      "pubmed21n0394.xml\n",
      "pubmed21n0938.xml\n",
      "pubmed21n0873.xml\n",
      "pubmed21n0009.xml\n",
      "pubmed21n0541.xml\n",
      "pubmed21n1040.xml\n",
      "pubmed21n0052.xml\n",
      "pubmed21n0442.xml\n",
      "pubmed21n0645.xml\n",
      "pubmed21n0710.xml\n",
      "pubmed21n0768.xml\n",
      "pubmed21n0661.xml\n",
      "pubmed21n0751.xml\n",
      "pubmed21n0051.xml\n",
      "pubmed21n0136.xml\n",
      "pubmed21n1060.xml\n",
      "pubmed21n0086.xml\n",
      "pubmed21n0975.xml\n",
      "pubmed21n0999.xml\n",
      "pubmed21n0948.xml\n",
      "pubmed21n0835.xml\n",
      "pubmed21n0850.xml\n",
      "pubmed21n0959.xml\n",
      "pubmed21n0781.xml\n",
      "pubmed21n0673.xml\n",
      "pubmed21n0407.xml\n",
      "pubmed21n0565.xml\n",
      "pubmed21n0923.xml\n",
      "pubmed21n0849.xml\n",
      "pubmed21n0943.xml\n",
      "pubmed21n0910.xml\n",
      "pubmed21n0516.xml\n",
      "pubmed21n0059.xml\n",
      "pubmed21n0037.xml\n",
      "pubmed21n0249.xml\n",
      "pubmed21n0988.xml\n",
      "pubmed21n0218.xml\n",
      "pubmed21n0075.xml\n",
      "pubmed21n0587.xml\n",
      "pubmed21n1025.xml\n",
      "pubmed21n0374.xml\n",
      "pubmed21n0518.xml\n",
      "pubmed21n0257.xml\n",
      "pubmed21n0570.xml\n",
      "pubmed21n0970.xml\n",
      "pubmed21n0979.xml\n",
      "pubmed21n0267.xml\n",
      "pubmed21n0680.xml\n",
      "pubmed21n1035.xml\n",
      "pubmed21n0157.xml\n",
      "pubmed21n0799.xml\n",
      "pubmed21n0354.xml\n",
      "pubmed21n1057.xml\n",
      "pubmed21n0912.xml\n",
      "pubmed21n0644.xml\n",
      "pubmed21n0438.xml\n",
      "pubmed21n0408.xml\n",
      "pubmed21n1059.xml\n",
      "pubmed21n0110.xml\n",
      "pubmed21n1033.xml\n",
      "pubmed21n0271.xml\n",
      "pubmed21n0306.xml\n",
      "pubmed21n0620.xml\n",
      "pubmed21n0898.xml\n",
      "pubmed21n0656.xml\n",
      "pubmed21n0020.xml\n",
      "pubmed21n0601.xml\n",
      "pubmed21n0648.xml\n",
      "pubmed21n0689.xml\n",
      "pubmed21n0834.xml\n",
      "pubmed21n0379.xml\n",
      "pubmed21n0800.xml\n",
      "pubmed21n0434.xml\n",
      "pubmed21n0011.xml\n",
      "pubmed21n0090.xml\n",
      "pubmed21n0214.xml\n",
      "pubmed21n0863.xml\n",
      "pubmed21n0889.xml\n",
      "pubmed21n1017.xml\n",
      "pubmed21n0096.xml\n",
      "pubmed21n0749.xml\n",
      "pubmed21n0501.xml\n",
      "pubmed21n0961.xml\n",
      "pubmed21n0072.xml\n",
      "pubmed21n0300.xml\n",
      "pubmed21n0694.xml\n",
      "pubmed21n0080.xml\n",
      "pubmed21n0992.xml\n",
      "pubmed21n0753.xml\n",
      "pubmed21n0076.xml\n",
      "pubmed21n0698.xml\n",
      "pubmed21n0299.xml\n",
      "pubmed21n0233.xml\n",
      "pubmed21n0101.xml\n",
      "pubmed21n0219.xml\n",
      "pubmed21n0411.xml\n",
      "pubmed21n0825.xml\n",
      "pubmed21n0657.xml\n",
      "pubmed21n0509.xml\n",
      "pubmed21n0494.xml\n",
      "pubmed21n0774.xml\n",
      "pubmed21n0893.xml\n",
      "pubmed21n0399.xml\n",
      "pubmed21n0180.xml\n",
      "pubmed21n0065.xml\n",
      "pubmed21n0357.xml\n",
      "pubmed21n0757.xml\n",
      "pubmed21n1047.xml\n",
      "pubmed21n0511.xml\n",
      "pubmed21n0467.xml\n",
      "pubmed21n1055.xml\n",
      "pubmed21n0003.xml\n",
      "pubmed21n0978.xml\n",
      "pubmed21n0282.xml\n",
      "pubmed21n0148.xml\n",
      "pubmed21n0826.xml\n",
      "pubmed21n0904.xml\n",
      "pubmed21n0122.xml\n",
      "pubmed21n0647.xml\n",
      "pubmed21n0784.xml\n",
      "pubmed21n0253.xml\n",
      "pubmed21n0387.xml\n",
      "pubmed21n0827.xml\n",
      "pubmed21n0064.xml\n",
      "pubmed21n0684.xml\n",
      "pubmed21n0921.xml\n",
      "pubmed21n0435.xml\n",
      "pubmed21n0174.xml\n",
      "pubmed21n0151.xml\n",
      "pubmed21n0634.xml\n",
      "pubmed21n0736.xml\n",
      "pubmed21n0058.xml\n",
      "pubmed21n0288.xml\n",
      "pubmed21n0091.xml\n",
      "pubmed21n0924.xml\n",
      "pubmed21n0833.xml\n",
      "pubmed21n0185.xml\n",
      "pubmed21n0724.xml\n",
      "pubmed21n0880.xml\n",
      "pubmed21n0792.xml\n",
      "pubmed21n0269.xml\n",
      "pubmed21n0847.xml\n",
      "pubmed21n0823.xml\n",
      "pubmed21n0574.xml\n",
      "pubmed21n0262.xml\n",
      "pubmed21n0513.xml\n",
      "pubmed21n0936.xml\n",
      "pubmed21n0577.xml\n",
      "pubmed21n0563.xml\n",
      "pubmed21n0410.xml\n",
      "pubmed21n0470.xml\n",
      "pubmed21n0719.xml\n",
      "pubmed21n0770.xml\n",
      "pubmed21n0447.xml\n",
      "pubmed21n0198.xml\n",
      "pubmed21n0573.xml\n",
      "pubmed21n0886.xml\n",
      "pubmed21n0181.xml\n",
      "pubmed21n0639.xml\n",
      "pubmed21n0454.xml\n",
      "pubmed21n0883.xml\n",
      "pubmed21n0527.xml\n",
      "pubmed21n0457.xml\n",
      "pubmed21n0261.xml\n",
      "pubmed21n0500.xml\n",
      "pubmed21n1031.xml\n",
      "pubmed21n0637.xml\n",
      "pubmed21n0414.xml\n",
      "pubmed21n0215.xml\n",
      "pubmed21n0755.xml\n",
      "pubmed21n0193.xml\n",
      "pubmed21n0134.xml\n",
      "pubmed21n0167.xml\n",
      "pubmed21n0263.xml\n",
      "pubmed21n0532.xml\n",
      "pubmed21n0659.xml\n",
      "pubmed21n0405.xml\n",
      "pubmed21n0243.xml\n",
      "pubmed21n0785.xml\n",
      "pubmed21n0456.xml\n",
      "pubmed21n0820.xml\n",
      "pubmed21n0888.xml\n",
      "pubmed21n0848.xml\n",
      "pubmed21n0490.xml\n",
      "pubmed21n0677.xml\n",
      "pubmed21n0445.xml\n",
      "pubmed21n0798.xml\n",
      "pubmed21n0311.xml\n",
      "pubmed21n0419.xml\n",
      "pubmed21n0383.xml\n",
      "pubmed21n0145.xml\n",
      "pubmed21n0159.xml\n",
      "pubmed21n0266.xml\n",
      "pubmed21n0258.xml\n",
      "pubmed21n0016.xml\n",
      "pubmed21n1053.xml\n",
      "pubmed21n0337.xml\n",
      "pubmed21n0325.xml\n",
      "pubmed21n0851.xml\n",
      "pubmed21n0461.xml\n",
      "pubmed21n0548.xml\n",
      "pubmed21n0593.xml\n",
      "pubmed21n0232.xml\n",
      "pubmed21n0764.xml\n",
      "pubmed21n0239.xml\n",
      "pubmed21n0452.xml\n",
      "pubmed21n0480.xml\n",
      "pubmed21n0712.xml\n",
      "pubmed21n0487.xml\n",
      "pubmed21n0591.xml\n",
      "pubmed21n0451.xml\n",
      "pubmed21n0265.xml\n",
      "pubmed21n0108.xml\n",
      "pubmed21n0123.xml\n",
      "pubmed21n0236.xml\n",
      "pubmed21n0592.xml\n",
      "pubmed21n0192.xml\n",
      "pubmed21n0697.xml\n",
      "pubmed21n0493.xml\n",
      "pubmed21n0919.xml\n",
      "pubmed21n0981.xml\n",
      "pubmed21n0602.xml\n",
      "pubmed21n0417.xml\n",
      "pubmed21n0488.xml\n",
      "pubmed21n0545.xml\n",
      "pubmed21n0046.xml\n",
      "pubmed21n1003.xml\n",
      "pubmed21n0990.xml\n",
      "pubmed21n0107.xml\n",
      "pubmed21n0726.xml\n",
      "pubmed21n1014.xml\n",
      "pubmed21n0598.xml\n",
      "pubmed21n0742.xml\n",
      "pubmed21n1052.xml\n",
      "pubmed21n0776.xml\n",
      "pubmed21n0870.xml\n",
      "pubmed21n0235.xml\n",
      "pubmed21n0433.xml\n",
      "pubmed21n1056.xml\n",
      "pubmed21n0465.xml\n",
      "pubmed21n0498.xml\n",
      "pubmed21n0460.xml\n",
      "pubmed21n0323.xml\n",
      "--- 609.4979477246602 minutes ---\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "path=\"data\"\n",
    "\n",
    "all_files_df=import_all_files(path, order_files=True)\n",
    "\n",
    "#save results\n",
    "all_files_df.to_pickle(\"variables/all_files_df\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 31820051 papers\n"
     ]
    }
   ],
   "source": [
    "print('There are {} papers'.format(all_files_df.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filtering abstracts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### First filtering by: \n",
    "- Empty abstracts\n",
    "- Language\n",
    "- Correction abstracts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After first cleaning, there are 19511885 papers\n",
      "--- 0.5490604281425476 minutes ---\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Eliminate empty abstracts\n",
    "clean_df=all_files_df[all_files_df.AbstractText != '']\n",
    "\n",
    "# Eliminate non-english papers\n",
    "clean_df=clean_df[clean_df.Language == 'eng']\n",
    "\n",
    "# Eliminate correction abstracts\n",
    "weird_text='An amendment to this paper has been published and can be accessed via a link at the top of the paper.'\n",
    "clean_df=clean_df[clean_df.AbstractText != weird_text]\n",
    "\n",
    "# print size\n",
    "print('After first cleaning, there are {} papers'.format(clean_df.shape[0]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Threshold and cut off\n",
    "Based on the abstract string length. We observed that papers with either too short or too long abstracts corresponded to non-meaningful abstracts. Examples of non-meaningful abstracts are texts containing only links, symbols, or simply empty. Those type of abstracts create artifacts in the embedding that hinder the quality of the visualization, therefore they were removed.\n",
    "\n",
    "The values of cutoff and threshold were chosen trying to balance the fraction of papers filtered out and the artifacts observed in the embedding."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cut off = 4000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before cut off, there are 19511885 papers\n",
      "After cut off, there are 19500414 papers\n"
     ]
    }
   ],
   "source": [
    "print('Before cut off, there are {} papers'.format(clean_df.shape[0]))\n",
    "abstracts=clean_df['AbstractText'].tolist()\n",
    "len_strings=map(len, abstracts)\n",
    "len_abstracts=np.fromiter(len_strings, dtype=np.int64,count=len(abstracts))\n",
    "\n",
    "\n",
    "cut_off=4000\n",
    "clean_df=clean_df[len_abstracts<cut_off]\n",
    "print('After cut off, there are {} papers'.format(clean_df.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Threshold = 250"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After threshold, there are 19016308 papers\n"
     ]
    }
   ],
   "source": [
    "threshold=250\n",
    "len_short_abstracts=len_abstracts[len_abstracts<cut_off]\n",
    "clean_df=clean_df[len_short_abstracts>threshold]\n",
    "print('After threshold, there are {} papers'.format(clean_df.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save df\n",
    "clean_df.to_pickle(\"variables/clean_df\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.0 ('analytics')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  },
  "vscode": {
   "interpreter": {
    "hash": "f4203a5cb41a9b3ea84e99fe66c07097d8c9544709fbc84b9fa75e017137c69a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
